jetty-proxy/src/test/java/org/eclipse/jetty/proxy/ProxyServletTest.java

```diff
@@ -549,6 +612,79 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
+            {
+                PrintWriter writer = resp.getWriter();
+                writer.write(req.getHeader("Via"));
+                writer.flush();
+            }
+        });
+        String viaHost = "my-good-via-host.example.org";
+        startProxy(derivedProxyServlet, Collections.singletonMap("viaHost", viaHost));
+        startClient();
+
+        HttpRequest proxyRequest = mockProxyRequest();
+        derivedProxyServlet.addViaHeader(proxyRequest);
+
+        ContentResponse response = client.GET("http://localhost:" + serverConnector.getLocalPort());
+        String expectedVia = protocol + " " + viaHost;
+
+        assertThat("Response expected to contain content of Via Header from the request",
+                   response.getContentAsString(),
+                   Matchers.equalTo(expectedVia));
+    }
+
+    @ParameterizedTest
+    @MethodSource("impls")
+    public void testProxyViaHeaderIsPresent(Class<? extends ProxyServlet> proxyServletClass) throws Exception
+    {
+        startServer(new HttpServlet()
+        {
+            @Override
+            protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException
+            {
+                PrintWriter writer = resp.getWriter();
+                writer.write(req.getHeader("Via"));
+                writer.flush();
+            }
+        });
+        String viaHost = "my-good-via-host.example.org";
+        startProxy(proxyServletClass, Collections.singletonMap("viaHost", viaHost));
+        startClient();
+
+        ContentResponse response = client.GET("http://localhost:" + serverConnector.getLocalPort());
+        assertThat("Response expected to contain content of Via Header from the request",
+                   response.getContentAsString(),
+                   Matchers.equalTo("1.1 " + viaHost));
+    }
+
+    @ParameterizedTest
+    @MethodSource("implsWithProtocols")
+    public void testProxyViaHeaderForVariousProtocols(Class<? extends ProxyServlet> proxyServletClass, String protocol) throws Exception
+    {
+        AbstractProxyServlet proxyServlet = proxyServletClass.getDeclaredConstructor().newInstance();
+        String host = InetAddress.getLocalHost().getHostName();
+        HttpServletRequest clientRequest = mockClientRequest(protocol);
+        HttpRequest proxyRequest = mockProxyRequest();
+
+        proxyServlet.addViaHeader(clientRequest, proxyRequest);
+
+        String expectedProtocol = protocol.startsWith("HTTP") ? protocol.split("/", 2)[1] : protocol;
+        String expectedVia = expectedProtocol + " " + host;
+        String expectedViaWithLocalhost = expectedProtocol + " localhost";
+
+        assertThat("Response expected to contain a Via header with the right protocol version and host",
+                   proxyRequest.getHeaders().getField("Via").getValue(),
+                   Matchers.anyOf(Matchers.equalTo(expectedVia), Matchers.equalTo(expectedViaWithLocalhost)));
+    }
```

Comment:
Please no.
This is not testing `ProxyServlet`, it is just testing that this test calls `addViaHeader()`, which well... it does.

If by mistake the real call to `addViaHeader()` is removed from `ProxyServlet` (e.g. a bad merge or some mistake), this test will pass, but `ProxyServlet` won't work.

Reply:
Isn't it testing that `addViaHeader` actually does add the Via header correctly for all the protocols?
I think this test to check the method in isolation is fine.  OK it is also tested by the previous test that checks the whole servlet, but testing the method directly is OK as well... so long as it doesn't need a third party mock library to do so.

Reply:
Ya, that's what it's doing @gregw . I agree that it's helpful in that sense.